agent/cache/cache.go

```diff
@@ -819,14 +847,22 @@ func (c *Cache) fetch(key string, r getOptions, allowNew bool, attempt uint, ign
 				return
 			}
 
+			var totalWait time.Duration
+
 			// If we're over the attempt minimum, start an exponential backoff.
 			if wait := backOffWait(attempt); wait > 0 {
-				time.Sleep(wait)
+				totalWait += wait
 			}
 
 			// If we have a timer, wait for it
 			if tEntry.Opts.RefreshTimer > 0 {
-				time.Sleep(tEntry.Opts.RefreshTimer)
+				totalWait += tEntry.Opts.RefreshTimer
+			}
+
+			select {
+			case <-time.After(totalWait):
+			case <-handle.stopCh:
+				return
```

Comment:
What happens if another `fetch` happens in another goroutine right after this point but before our own `fetch` call?

 * other `fetch` will see our handle, close our chan (no effect) and replace it with their own with a greater ID (say 10)
 * other `fetch` will then spawn `fetch.internal` and start an RPC 
 * we continue into our own `fetch` see the handle with ID 10 and close it's chan, make our new one with ID 11, and then re-enter `fetch.internal`
 * other `fetch.internal` returns from RPC, see's it's chan closed and exits (wasted RPC)
 * we complete our RPC and assuming no new one has started, update the cache entry.

**Conclusion:** seems safe although it does mean we still end up doing some redundant RPCs. I think that's OK as a quick fix.

Hmmm so one thing this could possible cause would be effectively a DOS of servers on error. Normally we'd have the retry backoff protecting servers if repeated errors happen. It would look like this:

 1. `fetch.internal 1` gets an error, updates entry, starts backoff wait
 2. client loop has a smaller backoff (or none) so retries which results in a new `fetch 2`
 3. `fetch 2` will see `Fetching == false` and so spawn a `fetch.internal 2` while closing the stop chan for `fetch.internal 1`. it will _immediately_ make another request to servers even though we are supposed to be waiting.
 4. eventually `fetch.internal 1` finishes it's backoff wait and see's it's been deposed and so just returns
 5. `fetch.internal 2` also hits error, starts backoff
 6. client retries sooner again etc.

We _might_ be OK because our internal cache clients generally use `Notify` which _probably_ does the right thing in that it does its own backoff loop IIRC. **But this one seems at least a bit risky that this fix potentially means that we'll bypass error backoffs?**

Reply:
> Hmmm so one thing this could possible cause would be effectively a DOS of servers on error. 
That's actually already happening on error with the existing code, since getting an error is `Fetching=false`.